File: Mooga/src/de/lumpn/mooga/selection/BinaryTournamentSelection.java
package de.lumpn.mooga.selection;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import de.lumpn.mooga.Individual;

public final class BinaryTournamentSelection implements Selection {

	public BinaryTournamentSelection(Random random) {
		this.random = random;
	}

	@Override
	public Individual select(List<Individual> individuals) {
		int size = individuals.size();
		int pos1 = random.nextInt(size);
		int pos2 = random.nextInt(size);
		return individuals.get(Math.min(pos1, pos2));
	}

	@Override
	public List<Individual> select(List<Individual> individuals, int count) {
		List<Individual> result = new ArrayList<Individual>();
		while (result.size() < count) {
			Individual individual = select(individuals);
			if (!result.contains(individual)) {
				result.add(individual);
			}
		}
		return result;
	}

	private final Random random;
}

File: csharp/InfiniteZelda/Mooga/BinaryTournamentSelection.cs
using System;
using System.Collections.Generic;

namespace Lumpn.Mooga
{
    public sealed class BinaryTournamentSelection : ISelection
    {
        public BinaryTournamentSelection(IRandom random)
        {
            this.random = random;
        }

        public IIndividual Select(List<IIndividual> individuals)
        {
            int size = individuals.Count;
            int pos1 = random.NextInt(size);
            int pos2 = random.NextInt(size);
            int pos = Math.Min(pos1, pos2);
            return individuals[pos];
        }

        public List<IIndividual> Select(List<IIndividual> individuals, int count)
        {
            Debug.Assert(count < individuals.Count);

            var result = new List<IIndividual>();
            while (result.Count < count)
            {
                var individual = Select(individuals);
                if (!result.Contains(individual))
                {
                    result.Add(individual);
                }
            }
            return result;
        }

        private readonly IRandom random;
    }
}